src/src/Helpers/TarStream.cpp

```diff
@@ -0,0 +1,603 @@
+          }
+        } else {
+          memset(&_tarHeader, 0, TAR_HEADER_SIZE);
+          _streamState = TarStreamState_e::ReadingFinal;
+          # if TAR_STREAM_DEBUG
+          addLog(LOG_LEVEL_INFO, concat(F("TarStream: Switch from "), F("ReadingSlack to ReadingFinal")));
+          # endif // if TAR_STREAM_DEBUG
+        }
+        _headerPosition = 0;
+        result          = _tarData[_headerPosition];
+      }
+      break;
+    }
+    case TarStreamState_e::ReadingFinal:
+    {
+      _headerPosition++;
+      result = 0;
+
+      if (_headerPosition == (TAR_BLOCK_SIZE * 2)) {
+        result = EOF;
+        # if TAR_STREAM_DEBUG
+        addLog(LOG_LEVEL_INFO, concat(F("TarStream: "), F("Reached ReadingFinal EOF")));
+        # endif // if TAR_STREAM_DEBUG
+      } else if (_headerPosition > (TAR_BLOCK_SIZE * 2)) {
+        result       = EOF;
+        _streamState = TarStreamState_e::Error;
+        # if TAR_STREAM_DEBUG
+        addLog(LOG_LEVEL_INFO, concat(F("TarStream: Switch from "), F("ReadingFinal to Error")));
+        # endif // if TAR_STREAM_DEBUG
+      }
+      break;
+    }
+    case TarStreamState_e::WritingHeader: // Not here
+    case TarStreamState_e::WritingFile:
+    case TarStreamState_e::WritingSlack:
+    case TarStreamState_e::WritingFinal:
+    case TarStreamState_e::WritingDone:
+      break;
+    case TarStreamState_e::Error:
+      result = EOF; // Endstate, something went wrong
+      break;
+  }
+
+  if (_tarRemaining > 0u) {
+    _tarRemaining--;
+  }
+
+  return result;
+}
+
+int TarStream::peek() {
+  int result = 0;
+
+  # if TAR_STREAM_PEEK
+
+  switch (_streamState) {
+    case TarStreamState_e::Initial:
+      break;
+    case TarStreamState_e::ReadingHeader:
+    {
+      if (_headerPosition < TAR_HEADER_SIZE - 1) {
+        result = _tarData[_headerPosition];
+      } else if (_headerPosition < TAR_BLOCK_SIZE - 1) {
+        result = 0;
+      } else if (_currentFile) {
+        result = _currentFile.peek();
+      }
+      break;
+    }
+    case TarStreamState_e::ReadingFile:
+    {
+      if (_tarPosition < _currentIndex.fileSize + 1) {
+        result = _currentFile.peek();
+      }
+      break;
+    }
+    case TarStreamState_e::ReadingFinal:
+    {
+      if (_headerPosition == (TAR_BLOCK_SIZE * 2) - 1) {
+        result = EOF;
+      }
+      break;
+    }
+    case TarStreamState_e::ReadingSlack:
+    case TarStreamState_e::WritingHeader: // Ignore
+    case TarStreamState_e::WritingFile:
+    case TarStreamState_e::WritingSlack:
+    case TarStreamState_e::WritingFinal:
+    case TarStreamState_e::WritingDone:
+      break;
+    case TarStreamState_e::Error:
+      result = EOF; // Endstate, something went wrong
+      break;
+  }
+  # endif // if TAR_STREAM_PEEK
+
+  return result;
+}
+
+void TarStream::flush() {
+  if (_currentFile) {
+    _currentFile.flush();
+  }
+  _tarRemaining = 0u;
+}
+
+size_t TarStream::size() {
+  return _tarSize;
+}
+
+const char * TarStream::name() {
```

Comment:
This is a const function.

Reply:
It's an volatile/override method, so must have the same signature as the base class...